File: _3_Learning/_02_Algorithms/Week2/RandomizedQueue.java
import edu.princeton.cs.algs4.StdRandom;

import java.util.Iterator;
import java.util.NoSuchElementException;

public class RandomizedQueue<Item> implements Iterable<Item> {

    private Item[] items;
    private int n;

    // construct an empty randomized queue
    public RandomizedQueue() {
        items = (Item[]) new Object[2];
    }

    // is the randomized queue empty?
    public boolean isEmpty() {
        return n == 0;
    }

    // return the number of items on the randomized queue
    public int size() {
        return n;
    }

    private void resize(int capacity) {
        Item[] copy = (Item[]) new Object[capacity];
        for (int i = 0; i < n; i++)
            copy[i] = items[i];
        items = copy;
    }

    // add the item
    public void enqueue(Item item) {
        if (null == item) throw new IllegalArgumentException();
        if (n == items.length) resize(n * 2);
        items[n++] = item;

    }

    // remove and return a random item
    public Item dequeue() {
        if (isEmpty()) throw new IllegalArgumentException();
        int index = StdRandom.uniform(n);
        Item item = items[index];
        if (index != n - 1)
            items[index] = items[n - 1];
        items[n - 1] = null;
        n--;
        if (n < ((items.length) / 4))
            resize(items.length / 2);
        return item;

    }

    // return a random item (but do not remove it)
    public Item sample() {
        if (isEmpty()) throw new IllegalArgumentException();
        return items[StdRandom.uniform(n)];
    }

    // return an independent iterator over items in random order
    public Iterator<Item> iterator() {
        return new RandomizedQueueIterator();
    }

    private class RandomizedQueueIterator implements Iterator<Item> {
        private Item[] rdq;
        private int current;

        public RandomizedQueueIterator() {

            rdq = (Item[]) new Object[n];
            current = 0;

            for (int k = 0; k < n; k++) {
                rdq[k] = items[k];
            }

            StdRandom.shuffle(rdq);
        }

        public boolean hasNext() {
            return current < n;
        }

        public void remove() {
            throw new UnsupportedOperationException();
        }

        public Item next() {

            if (!hasNext()) throw new NoSuchElementException();
            Item it = rdq[current];
            current++;
            return it;
        }

    }

    // unit testing (required)
    public static void main(String[] args) {
        RandomizedQueue<String> rq = new RandomizedQueue();
        rq.enqueue("AAA");
        System.out.println(rq.isEmpty());
    }

}

File: _3_Learning/_02_Algorithms/Week2/RandomizedQueue.py
from random import randint
import sys
class RandomizedQueue():
    def __init__(self):
        self.items = [None] * 2
        self.n = 0

    def isEmpty(self):
        return self.n==0

    def size(self):
        return self.n

    def resize(self,capacity):
        if not isinstance(capacity,int):
            raise IllegalArgumentException
        copy = [None]*capacity
        for i in range(self.n):
            copy[i] = self.items[i]
        self.items = copy

    def enqueue(self,item):
        # double size when array is full
        if not item:
            raise IllegalArgumentException
        if self.n==len(self.items):
            self.resize(self.n*2)
        self.items[self.n] = item
        self.n+=1

    def dequeue(self):
        # halve size when array is 1/4 full
        if(self.isEmpty()):
            raise NoSuchElementException
        index = randint(0,self.n)
        item = self.items[index]
        if index!=self.n-1:
            self.items[index] = self.items[self.n-1]  # exchange items[index] and the last item
        self.items[self.n-1] = None
        self.n -= 1
        if self.n < (len(self.items)/4):
            self.resize(int(len(self.items)/2))
        return item

    def sample(self):
        if(self.isEmpty()):
            raise NoSuchElementException()
        return self.items[randint(0,self.n)]

    def iterator(self):
        return RandomizedQueueIterator(self)

class RandomizedQueueIterator():
    def __init__(self,rq):
        self.rdq = rq.items
        self.current = 0

    def hasNext(self):
        return self.current < len(self.rdq)

    def remove(self):
        raise UnsupportedOperationException

    def next(self):
        if not self.hasNext():
            raise NoSuchElementException
        item = self.rdq[self.current]
        self.current +=1
        return item

class IllegalArgumentException(Exception):
    pass
class NoSuchElementException(Exception):
    pass
class UnsupportedOperationException(Exception):
    pass

if __name__ == '__main__':
    with open(sys.argv[1]) as f:
        data = f.readlines()
        rq = RandomizedQueue()
        for i in data:
            rq.enqueue(i)
        for j in range(int(sys.argv[2])):
            print(rq.dequeue())
